include/oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/for_each.hpp

```diff
@@ -0,0 +1,94 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/sycl.hpp>
+
+#include "../../detail/sycl_utils.hpp"
+#include "../detail.hpp"
+#include "../init.hpp"
+#include "../util.hpp"
+#include "../zip_view.hpp"
+#include "execution_policy.hpp"
+
+namespace oneapi::dpl::experimental::dr::sp
+{
+
+template <typename ExecutionPolicy, distributed_range R, typename Fn>
```

Comment:
1) What kinds of execution policy are supported? (by design and by this implementation)
As far the C++20 concepts (`distributed_range` at least) are already used, it makes sense to add constrains on an execution policy as well.
In oneDPL, there is `oneapi::dpl::is_execution_policy_v` for example.

2) Does it make sense to add a constrain on `Fn` as well?

Reply:
1) The first line in the method is static_assert that the only supported execution policy is
```
static_assert( // currently only one policy supported
        std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, sycl_device_collection>);
```
IMO this is enough and simple

2) I don't think so. There are no constrains for functor in [std::for_each](https://en.cppreference.com/w/cpp/algorithm/for_each)

Reply:
1. Having "concepts"  for some parameter(s) in a signature and having "static_assert" for the other parameter(s) - not consistent approach. Or we use type requirements (concepts and contains) in a algo signature or  don't use at all. As far as we use C++20, I would vote for "concepts and constrains" usage. A user of oneDPL usually sees into the documentation where a signature there is.
2. But there is in [std::ranges::for_each](https://en.cppreference.com/w/cpp/algorithm/ranges/for_each)

Reply:
Can you explain what's problematic about using the name `par_unseq` as a catch-all execution policy to execute across multiple devices?

From my understanding of the spec, [`par_unseq` requires](https://eel.is/c++draft/algorithms.parallel#exec-7):

- Execution in "unordered fashion" in "unspecified threads of execution," and
- These threads must provide "weakly parallel forward progress guarantees."

From what I understand, threads in SYCL have weakly parallel forward progress guarantees.  Is it the "threads of execution" part that would be violated?

Other libraries [like Nvidia's stdpar](https://developer.nvidia.com/blog/accelerating-standard-c-with-gpus-using-stdpar/) use `par_unseq` to mean "execute in parallel, including on GPUs if that's where the data is."  This seems like a reasonable choice to me.  Is there a convention in oneDPL that `par_unseq` is CPU only, or is there another good reason to not use `par_unseq` as the "sane default" that includes potentially running on a GPU if that's where the data is resident?